File: ios/LookSeeProto/Models/BusinessMediaHistoryModels.swift
//
//  BusinessMediaHistoryModels.swift
//  LookSeeProto
//
//  Models returned by the landmark media-history API.
//

import Foundation

struct BusinessMediaHistoryResponse: Decodable {
    let landmarkId: String
    let landmarkLabel: String
    let items: [BusinessMediaHistoryItem]
    let count: Int
    let nextToken: String?
}

struct BusinessMediaHistoryUploader: Decodable, Hashable {
    let displayName: String?
    let email: String?
    let userId: String?

    var displayText: String {
        for candidate in [displayName, email, userId] {
            let cleaned = candidate?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            if !cleaned.isEmpty {
                return cleaned
            }
        }
        return "Unknown uploader"
    }
}

enum BusinessMediaLifecycleState: String, Hashable {
    case processing
    case ready
    case failed
    case unknown

    init(backendValue: String?) {
        let normalized = Self.normalize(backendValue)

        switch normalized {
        case "ready", "complete", "completed", "success", "succeeded":
            self = .ready
        case "processing", "retrying", "pending", "upload pending", "initiated", "uploaded":
            self = .processing
        case "failed", "failure", "error", "rejected":
            self = .failed
        default:
            self = .unknown
        }
    }

    var displayTitle: String {
        switch self {
        case .processing: return "Processing"
        case .ready: return "Ready"
        case .failed: return "Failed"
        case .unknown: return "Unknown"
        }
    }

    private static func normalize(_ value: String?) -> String {
        value?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "_", with: " ")
            .replacingOccurrences(of: "-", with: " ")
            .lowercased() ?? ""
    }
}

enum BusinessMediaHistoryRole: Hashable {
    case positive
    case hardNegative
    case unknown(String)

    init(backendValue: String) {
        let cleaned = backendValue
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
        let compact = cleaned
            .replacingOccurrences(of: "_", with: "")
            .replacingOccurrences(of: "-", with: "")
            .replacingOccurrences(of: " ", with: "")

        switch compact {
        case "positive": self = .positive
        case "negative", "hardnegative": self = .hardNegative
        default: self = .unknown(cleaned)
        }
    }

    var displayTitle: String {
        switch self {
        case .positive: return "Positive"
        case .hardNegative: return "Negative"
        case .unknown: return "Unknown role"
        }
    }
}

enum BusinessMediaHistoryKind: Hashable {
    case video
    case photo
    case unknown(String)

    init(backendValue: String) {
        let cleaned = backendValue
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()

        switch cleaned {
        case "video", "movie": self = .video
        case "photo", "image": self = .photo
        default: self = .unknown(cleaned)
        }
    }

    var displayTitle: String {
        switch self {
        case .video: return "Video"
        case .photo: return "Image"
        case .unknown: return "Unknown media"
        }
    }

    var systemImage: String {
        switch self {
        case .video: return "video.fill"
        case .photo: return "photo.fill"
        case .unknown: return "questionmark.square.dashed"
        }
    }
}

struct BusinessMediaHistoryItem: Decodable, Identifiable, Hashable {
    let id: String
    let submissionId: String
    let batchId: String?
    let datasetRole: String
    let mediaKind: String
    let originalFilename: String?
    let contentType: String?

    /// Canonical status returned by the newest history Lambda. Older API
    /// responses may still place their raw pipeline status here.
    let status: String?

    /// Original writer status retained by the history Lambda for diagnostics.
    let rawStatus: String?

    let uploadedBy: BusinessMediaHistoryUploader
    let uploadedAt: Int64?
    let uploadedAtISO: String?
    let thumbnailUrl: URL?
    let thumbnailSource: String?
    let retryCount: Int?
    let lastRetryAt: Int64?
    let failureReason: String?

    var role: BusinessMediaHistoryRole {
        BusinessMediaHistoryRole(backendValue: datasetRole)
    }

    var kind: BusinessMediaHistoryKind {
        BusinessMediaHistoryKind(backendValue: mediaKind)
    }

    var lifecycleState: BusinessMediaLifecycleState {
        BusinessMediaLifecycleState(backendValue: status ?? rawStatus)
    }

    var isPositive: Bool {
        role == .positive
    }

    var isVideo: Bool {
        kind == .video
    }

    var roleAndMediaTitle: String {
        "\(role.displayTitle) • \(kind.displayTitle)"
    }

    var mediaSystemImage: String {
        kind.systemImage
    }

    var normalizedStatus: String {
        lifecycleState.displayTitle
    }

    var displayStatus: String {
        isProcessingDelayed ? "Delayed" : lifecycleState.displayTitle
    }

    var backendStatusText: String {
        let value = rawStatus ?? status
        let cleaned = value?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return cleaned.isEmpty ? "unknown" : cleaned
    }

    var uploadDate: Date? {
        if let uploadedAt, uploadedAt > 0 {
            return Date(timeIntervalSince1970: TimeInterval(uploadedAt))
        }

        guard let uploadedAtISO, !uploadedAtISO.isEmpty else {
            return nil
        }
        return ISO8601DateFormatter().date(from: uploadedAtISO)
    }

    var isProcessingDelayed: Bool {
        guard lifecycleState == .processing,
              let activityDate = lastRetryDate ?? uploadDate else {
            return false
        }
        return Date().timeIntervalSince(activityDate) >= 60 * 60
    }

    var canRetryProcessing: Bool {
        guard role == .hardNegative,
              let batchId,
              !batchId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              !submissionId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return false
        }

        return lifecycleState == .failed || isProcessingDelayed
    }

    var lastRetryDate: Date? {
        guard let lastRetryAt, lastRetryAt > 0 else { return nil }
        return Date(timeIntervalSince1970: TimeInterval(lastRetryAt))
    }

    var displayFilename: String {
        let cleaned = originalFilename?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return cleaned.isEmpty ? "Unnamed media" : cleaned
    }
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/models/BusinessMediaHistoryModels.kt
package looksee.angelll.com.models

import java.time.Instant

data class BusinessMediaHistoryResponse(
    val landmarkId: String = "",
    val landmarkLabel: String = "",
    val items: List<BusinessMediaHistoryItem> = emptyList(),
    val count: Int = 0,
    val nextToken: String? = null,
)

data class BusinessMediaHistoryUploader(
    val displayName: String? = null,
    val email: String? = null,
    val userId: String? = null,
) {
    val displayText: String
        get() = sequenceOf(displayName, email, userId)
            .mapNotNull { it.nonBlank() }
            .firstOrNull()
            ?: "Unknown uploader"
}

enum class BusinessMediaLifecycleState(val displayTitle: String) {
    PROCESSING("Processing"),
    READY("Ready"),
    FAILED("Failed"),
    UNKNOWN("Unknown");

    companion object {
        fun fromBackend(value: String?): BusinessMediaLifecycleState =
            when (value.normalizedBackendValue()) {
                "ready", "complete", "completed", "success", "succeeded" -> READY
                "processing", "retrying", "pending", "upload pending", "initiated",
                "uploaded",
                -> PROCESSING
                "failed", "failure", "error", "rejected" -> FAILED
                else -> UNKNOWN
            }
    }
}

enum class BusinessMediaHistoryRole(val displayTitle: String) {
    POSITIVE("Positive"),
    HARD_NEGATIVE("Negative"),
    UNKNOWN("Unknown role");

    companion object {
        fun fromBackend(value: String): BusinessMediaHistoryRole = when (
            value.trim().lowercase()
                .replace("_", "")
                .replace("-", "")
                .replace(" ", "")
        ) {
            "positive" -> POSITIVE
            "negative", "hardnegative" -> HARD_NEGATIVE
            else -> UNKNOWN
        }
    }
}

enum class BusinessMediaHistoryKind(
    val displayTitle: String,
    val iconKey: String,
) {
    VIDEO("Video", "video"),
    PHOTO("Image", "photo"),
    UNKNOWN("Unknown media", "unknown");

    companion object {
        fun fromBackend(value: String): BusinessMediaHistoryKind = when (
            value.trim().lowercase()
        ) {
            "video", "movie" -> VIDEO
            "photo", "image" -> PHOTO
            else -> UNKNOWN
        }
    }
}

data class BusinessMediaHistoryItem(
    val id: String = "",
    val submissionId: String = "",
    val batchId: String? = null,
    val datasetRole: String = "",
    val mediaKind: String = "",
    val originalFilename: String? = null,
    val contentType: String? = null,
    /** Canonical lifecycle status from the newest history API. */
    val status: String? = null,
    /** Original writer status retained for diagnostics. */
    val rawStatus: String? = null,
    val uploadedBy: BusinessMediaHistoryUploader = BusinessMediaHistoryUploader(),
    val uploadedAt: Long? = null,
    val uploadedAtISO: String? = null,
    val thumbnailUrl: String? = null,
    val thumbnailSource: String? = null,
    val retryCount: Int? = null,
    val lastRetryAt: Long? = null,
    val failureReason: String? = null,
) {
    val role: BusinessMediaHistoryRole
        get() = BusinessMediaHistoryRole.fromBackend(datasetRole)

    val kind: BusinessMediaHistoryKind
        get() = BusinessMediaHistoryKind.fromBackend(mediaKind)

    val lifecycleState: BusinessMediaLifecycleState
        get() = BusinessMediaLifecycleState.fromBackend(status ?: rawStatus)

    val isPositive: Boolean
        get() = role == BusinessMediaHistoryRole.POSITIVE

    val isVideo: Boolean
        get() = kind == BusinessMediaHistoryKind.VIDEO

    val roleAndMediaTitle: String
        get() = "${role.displayTitle} • ${kind.displayTitle}"

    /** Compatibility aliases retained for already-translated UI code. */
    val roleTitle: String
        get() = role.displayTitle

    val mediaTitle: String
        get() = kind.displayTitle

    val mediaIconKey: String
        get() = kind.iconKey

    val normalizedStatus: String
        get() = lifecycleState.displayTitle

    val backendStatusText: String
        get() = (rawStatus ?: status).nonBlank() ?: "unknown"

    val uploadInstant: Instant?
        get() {
            uploadedAt?.takeIf { it > 0 }?.let { return Instant.ofEpochSecond(it) }
            return uploadedAtISO.nonBlank()?.let { value ->
                runCatching { Instant.parse(value) }.getOrNull()
            }
        }

    val lastRetryInstant: Instant?
        get() = lastRetryAt?.takeIf { it > 0 }?.let(Instant::ofEpochSecond)

    val isProcessingDelayed: Boolean
        get() = isProcessingDelayedAt(Instant.now())

    fun isProcessingDelayedAt(now: Instant): Boolean {
        if (lifecycleState != BusinessMediaLifecycleState.PROCESSING) return false
        val activity = lastRetryInstant ?: uploadInstant ?: return false
        return !now.isBefore(activity.plusSeconds(PROCESSING_DELAY_SECONDS))
    }

    val canRetryProcessing: Boolean
        get() = canRetryProcessingAt(Instant.now())

    fun canRetryProcessingAt(now: Instant): Boolean {
        if (role != BusinessMediaHistoryRole.HARD_NEGATIVE ||
            batchId.nonBlank() == null || submissionId.isBlank()
        ) {
            return false
        }
        return lifecycleState == BusinessMediaLifecycleState.FAILED ||
            isProcessingDelayedAt(now)
    }

    val displayStatus: String
        get() = if (isProcessingDelayed) "Delayed" else lifecycleState.displayTitle

    val displayFilename: String
        get() = originalFilename.nonBlank() ?: "Unnamed media"

    private companion object {
        const val PROCESSING_DELAY_SECONDS = 60L * 60L
    }
}

private fun String?.nonBlank(): String? = this?.trim()?.takeIf(String::isNotEmpty)

private fun String?.normalizedBackendValue(): String = this
    ?.trim()
    ?.replace('_', ' ')
    ?.replace('-', ' ')
    ?.lowercase()
    .orEmpty()
